### PR TITLE
Move client consensus parts out of primitives and into client/consensus/api

### PR DESCRIPTION
**Pull request by gilescope**
polkadot companion: https://github.com/paritytech/polkadot/pull/3456

### CODE REVIEW

File: bin/node/testing/src/bench.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ use crate::client::{Client, Backend};
 use crate::keyring::*;
 use sc_client_db::PruningMode;
 use sc_executor::{NativeExecutor, WasmExecutionMethod};
-use sp_consensus::{
+use sc_consensus_api::block_import::{
```
Comment (gilescope) [inline bin/node/testing/src/bench.rs:32]: We should keep that the same - I will re-export this from the crate so there's less changes.

**Comment by gilescope** [comment]
Happy to bikeshead any name for BlockImportStatus except BlockImportResult - it's not a Result and calling it that just trips the reader up.

File: client/consensus/api/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// our error-chain could potentially blow up otherwise
+#![recursion_limit="128"]
+
+#[macro_use] extern crate log;
```
Comment (seunlanlege) [inline client/consensus/api/src/lib.rs:4]: hmm, is this really neccessary?

**Review by seunlanlege — APPROVED** [review]
lgtm

**Comment by andresilva** [comment]
(please don't merge this without my review, which will come shortly)

File: client/consensus/api/src/lib.rs (hunk 2 above)
Comment (gilescope) [reply client/consensus/api/src/lib.rs:1]: Have removed it. Let's see if the companion build complains...

Comment (andresilva) [reply client/consensus/api/src/lib.rs:1]: Missing license header.

File: client/consensus/babe/src/lib.rs [hunk 3]
```diff
@@ -502,7 +506,12 @@ where
 
 	let (worker_tx, worker_rx) = channel(HANDLE_BUFFER_SIZE);
 
-	let answer_requests = answer_requests(worker_rx, config.0, client, babe_link.epoch_changes.clone());
+	let answer_requests = answer_requests(
+		worker_rx,
+		config.0,
+		client,
+		babe_link.epoch_changes.clone());
```
Comment (andresilva) [inline client/consensus/babe/src/lib.rs:513]:
```suggestion
		babe_link.epoch_changes.clone(),
	);
```

File: client/network/test/src/lib.rs [hunk 4]
```diff
@@ -618,9 +626,11 @@ impl<B: BlockT> Verifier<B> for VerifierAdapter<B> {
...
 		let hash = header.hash();
-		self.verifier.lock().await.verify(origin, header, justifications, body).await.map_err(|e| {
-			self.failed_verifications.lock().insert(hash, e.clone());
-			e
+		self.verifier.lock().await
+			.verify(origin, header, justifications, body).await
+			.map_err(|e| {
+				self.failed_verifications.lock().insert(hash, e.clone());
+				e
 		})
```
Comment (andresilva) [inline client/network/test/src/lib.rs:634]:
```suggestion
			})
```

File: client/network/test/src/lib.rs [hunk 5]
```diff
@@ -352,7 +356,11 @@ impl<D, B> Peer<D, B> where
 		headers_only: bool,
 		inform_sync_about_new_best_block: bool,
 		announce_block: bool,
-	) -> H256 where F: FnMut(BlockBuilder<Block, PeersFullClient, substrate_test_runtime_client::Backend>) -> Block {
+	) -> H256
+		where
+			F: FnMut(BlockBuilder<Block, PeersFullClient,
+				substrate_test_runtime_client::Backend>) -> Block
```
Comment (andresilva) [inline client/network/test/src/lib.rs:362]: This formatting is a bit weird :\

File: client/network/test/src/lib.rs [hunk 6]
```diff
@@ -662,7 +672,10 @@ pub struct FullPeerConfig {
 	pub sync_mode: SyncMode,
 }
 
-pub trait TestNetFactory: Sized where <Self::BlockImport as BlockImport<Block>>::Transaction: Send {
+pub trait TestNetFactory: Sized
```
Comment (andresilva) [inline client/network/test/src/lib.rs:675]: I think I'd drop this and the formatting changes below.

File: primitives/consensus/common/src/import_queue.rs [hunk 7]
```diff
@@ -15,272 +15,5 @@
...
-		import_block.state_action = StateAction::ExecuteIfPossible;
-	}
-
-	let imported = import_handle.import_block(import_block, cache).await;
-	if let Some(metrics) = metrics.as_ref() {
-		metrics.report_verification_and_import(started.elapsed());
-	}
-	import_handler(imported)
-}
+pub type CacheKeyId = [u8; 4];
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue.rs:19]: Let's move this to `block_import.rs` and remove this file.

File: client/service/src/client/client.rs [hunk 8]
```diff
@@ -987,11 +1006,13 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
...
 		if block == last_finalized {
-			warn!("Possible safety violation: attempted to re-finalize last finalized block {:?} ", last_finalized);
+			warn!("Possible safety violation: attempted to re-finalize last finalized block {:?} ",
+				last_finalized);
 			return Ok(());
 		}
 
-		let route_from_finalized = sp_blockchain::tree_route(self.backend.blockchain(), last_finalized, block)?;
+		let route_from_finalized = sp_blockchain::tree_route(
+			self.backend.blockchain(), last_finalized, block)?;
```
Comment (andresilva) [inline client/service/src/client/client.rs:1015]: IMO this is worse than before

File: client/consensus/api/src/import_queue.rs [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+		_number: NumberFor<B>,
+		_success: bool) {}
+
+	/// Request a justification for the given block.
+	fn request_justification(&mut self, _hash: &B::Hash, _number: NumberFor<B>) {}
+}
+
+/// Block import successful result.
+#[derive(Debug, PartialEq)]
+pub enum BlockImportStatus<N: std::fmt::Debug + PartialEq> {
```
Comment (andresilva) [inline client/consensus/api/src/import_queue.rs:151]: Fine with me, the only alternative I can come up with would be `BlockImportReceipt`.

**Review by andresilva — CHANGES_REQUESTED** [review]
I don't think creating a new crate is necessary and I think you should be able to just use `sc-consensus`. Some of the formatting changes are bit unnecessary IMO, in particular in code that wasn't changed in any way. I think there's still more stuff that can be moved like `SelectChain` and `BlockAnnounceValidator`. Overall changes LGTM, just commented with some nits.

**Comment by andresilva** [comment]
Also needs a companion PR to polkadot, some imports there are broken now.

File: primitives/consensus/common/src/import_queue.rs (hunk 7 above)
Comment (gilescope) [reply primitives/consensus/common/src/import_queue.rs:19]: Happy to do this but it's referenced by some other (client only) bits in primitives that I have to move out first so will do that in a separate PR.

File: client/service/src/client/client.rs (hunk 8 above)
Comment (gilescope) [reply client/service/src/client/client.rs:1015]: undone.

File: client/network/test/src/lib.rs (hunk 6 above)
Comment (gilescope) [reply client/network/test/src/lib.rs:675]: undone.

File: client/network/test/src/lib.rs (hunk 5 above)
Comment (gilescope) [reply client/network/test/src/lib.rs:362]: agreed - have put it in vertical arg style.

**Comment by gilescope** [comment]
There is a companion PR: https://github.com/paritytech/polkadot/pull/3456 (but it's my first time so do please let me know if anything is out of kilter).

**Comment by gilescope** [comment]
Sorry! Will stop rebasing and merge from master from now on.

File: bin/node/executor/Cargo.toml [hunk 10]
```diff
@@ -40,6 +40,8 @@ pallet-transaction-payment = { version = "4.0.0-dev", path = "../../../frame/tra
 pallet-treasury = { version = "4.0.0-dev", path = "../../../frame/treasury" }
 sp-application-crypto = { version = "4.0.0-dev", path = "../../../primitives/application-crypto" }
 sp-consensus-babe = { version = "0.10.0-dev", path = "../../../primitives/consensus/babe" }
+sp-consensus = { version = "0.10.0-dev", path = "../../../primitives/consensus/common" }
+sc-consensus = { version = "0.10.0-dev", path = "../../../client/consensus/common" }
```
Comment (expenses) [inline bin/node/executor/Cargo.toml:44]: Do these need to be here if they aren't used in the crate?

File: client/basic-authorship/Cargo.toml [hunk 11]
```diff
@@ -32,5 +32,6 @@ sc-proposer-metrics = { version = "0.9.0", path = "../proposer-metrics" }
 
 [dev-dependencies]
 sc-transaction-pool = { version = "4.0.0-dev", path = "../transaction-pool" }
+sc-consensus = { version = "0.10.0-dev", path = "../consensus/common" }
```
Comment (expenses) [inline client/basic-authorship/Cargo.toml:35]: Ditto

File: client/offchain/Cargo.toml [hunk 12]
```diff
@@ -44,6 +44,7 @@ sc-transaction-pool = { version = "4.0.0-dev", path = "../transaction-pool" }
 sc-transaction-pool-api = { version = "4.0.0-dev", path = "../transaction-pool/api" }
 sp-tracing = { version = "4.0.0-dev", path = "../../primitives/tracing" }
 sp-consensus = { version = "0.10.0-dev", path = "../../primitives/consensus/common" }
+sc-consensus = { version = "0.10.0-dev", path = "../../client/consensus/common" }
```
Comment (expenses) [inline client/offchain/Cargo.toml:47]: Ditto

File: client/transaction-pool/Cargo.toml [hunk 13]
```diff
@@ -40,6 +40,7 @@ retain_mut = "0.1.3"
 assert_matches = "1.3.0"
 hex = "0.4"
 sp-consensus = { version = "0.10.0-dev", path = "../../primitives/consensus/common" }
+sc-consensus = { version = "0.10.0-dev", path = "../../client/consensus/common" }
```
Comment (expenses) [inline client/transaction-pool/Cargo.toml:43]: Ditto

File: test-utils/runtime/Cargo.toml [hunk 14]
```diff
@@ -52,6 +52,8 @@ serde = { version = "1.0.126", optional = true, features = ["derive"] }
 [dev-dependencies]
 sc-block-builder = { version = "0.10.0-dev", path = "../../client/block-builder" }
 sc-executor = { version = "0.10.0-dev", path = "../../client/executor" }
+sc-consensus = { version = "0.10.0-dev", path = "../../client/consensus/common" }
+sp-consensus = { version = "0.10.0-dev", path = "../../primitives/consensus/common" }
```
Comment (expenses) [inline test-utils/runtime/Cargo.toml:56]: Ditto

```suggestion
```

**Review by expenses — COMMENTED** [review]
I think we might be able to do some import trimming here of `sp-consensus` and `sc-consensus`.

File: bin/node/executor/Cargo.toml (hunk 10 above)
Comment (gilescope) [reply bin/node/executor/Cargo.toml]: nope. Great spot.

**Review by expenses — APPROVED** [review]
I think this probably good to go. Needs a few more reviews first though.

File: primitives/consensus/common/src/import_queue.rs [hunk 15]
```diff
@@ -15,292 +15,5 @@
...
-	/// Skip block exection and state verification.
-	pub skip_execution: bool,
-	/// Re-validate existing block.
-	pub import_existing: bool,
-	/// Do not compute new state, but rather set it to the given set.
-	pub state: Option<ImportedState<B>>,
-}
-
 /// Type of keys in the blockchain cache that consensus module could use for its needs.
 pub type CacheKeyId = [u8; 4];
```
Comment (tomusdrw) [inline primitives/consensus/common/src/import_queue.rs:19]: If that's the only type left here, maybe move it to top-level `lib.rs`?

File: client/consensus/common/src/lib.rs [hunk 16]
```diff
@@ -18,7 +18,25 @@
 
 //! Collection of common consensus specific implementations
 
+#[macro_use]
+extern crate log;
```
Comment (tomusdrw) [inline client/consensus/common/src/lib.rs:22]: Is this really needed?

**Review by tomusdrw — APPROVED** [review]

File: client/consensus/common/src/lib.rs (hunk 16 above)
Comment (gilescope) [reply client/consensus/common/src/lib.rs:22]: log is used in these two files: https://github.com/gilescope/substrate/blob/2ba954e5c7db603ff395ca1d17bcabf61f7956da/client/consensus/common/src/import_queue/basic_queue.rs#L103 and https://github.com/gilescope/substrate/blob/2ba954e5c7db603ff395ca1d17bcabf61f7956da/client/consensus/common/src/import_queue.rs#L213

Comment (tomusdrw) [reply client/consensus/common/src/lib.rs:22]: Sure, but you don't need the obsolete `macro_use` and `extern crate` for this. Just `use log::debug;` or `log::debug!()`.

File: primitives/consensus/common/src/import_queue.rs (hunk 15 above)
Comment (gilescope) [reply primitives/consensus/common/src/import_queue.rs:19]: Happy to move it one level up. I'm tempted to leave it in the `import_queue` namespace for backward compatibility.

Comment (gilescope) [reply primitives/consensus/common/src/import_queue.rs:19]: actually I can't see it referenced in polkadot so will move it up a level.

File: docs/STYLE_GUIDE.md [hunk 17]
```diff
@@ -2,6 +2,9 @@
 title: Style Guide for Rust in Substrate
 ---
 
+Where possible these styles are enforced by settings in `rustfmt.toml` so if you run `cargo fmt` 
+then you will adhear to most of these style guidelines automatically.
```
Comment (andresilva) [inline docs/STYLE_GUIDE.md:6]:
```suggestion
then you will adhere to most of these style guidelines automatically.
```

File: client/consensus/common/src/lib.rs (hunk 16 above)
Comment (gilescope) [reply client/consensus/common/src/lib.rs]: updated imports.

**Review by andresilva — APPROVED** [review]
LGTM. Sorry for delay in review 🙏

**Comment by gilescope** [comment]
bot merge

**Pull request merged**